### PR TITLE
make sure *os.File is pollable so Close unblocks any pending Read

### PR DESCRIPTION
**Pull request by songgao**
`(*os.File).Fd()` converts a non-blocking fd to blocking, which likely is why calling `Close()` doesn't unblock `Read()`. I tried using `SyscallConn()` (introduced [here](https://github.com/golang/go/issues/24331)) on the `*os.File` returned from `os.OpenFile("/dev/net/tun"...)`, but it breaks `Read()`, likely because of the `ioctl`s that happen afterwards but not before,  as explained [here](https://github.com/golang/go/issues/22939#issuecomment-348408364).

So instead, just open `/dev/net/tun` with `syscall.Open()`, with `O_NONBLOCK` flag, do those `ioctl`s, then pass the fd into `os.NewFile()`. Fortunately since go1.11 `os.NewFile()` now put the resulting `*os.File` into non-blocking mode if the fd passed in is in non-blocking mode. As a result it'd be using the runtime poller and `Close()` can unblock any pending `Read()`.

@Gnouc thanks for the pointers and initiating the work. Your test from #22 is included in this branch, and I've added you to `CONTRIBUTORS`.

fixes https://github.com/songgao/water/issues/22, resolves https://github.com/songgao/water/pull/52

### CODE REVIEW

**Comment by songgao** [comment]
@Gnouc @dsnet @lixin9311 Mind taking a look when you get a chance? Thanks!

File: syscalls_linux.go [hunk 1]
```diff
@@ -31,50 +31,52 @@ func ioctl(fd uintptr, request uintptr, argp uintptr) error {
...
 		return nil, err
 	}
 
-	if err = setDeviceOptions(file.Fd(), config); err != nil {
+	if err = setDeviceOptions(fd, config); err != nil {
 		return nil, err
 	}
 
-	ifce = &Interface{isTAP: true, ReadWriteCloser: file, name: name}
+	ifce = &Interface{isTAP: true, ReadWriteCloser: os.NewFile(fd, "tun"), name: name}
```
Comment (Gnouc) [inline syscalls_linux.go]: Name should be `"tap"` instead of `"tun"`.

File: ipv4_darwin_test.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
+// +build linux darwin
```
Comment (Gnouc) [inline ipv4_darwin_test.go]: This build tag is redundant, because this file is only considered in darwin.

**Comment by Gnouc** [comment]
@songgao LGTM, just some minor comments.

File: syscalls_linux.go (hunk 1 above)
Comment (songgao) [reply syscalls_linux.go]: Oh this is actually `tun` I think. Although this is a tap interface, it's still from `/dev/net/tun`

**Pull request merged**